torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -743,6 +757,145 @@ c10::optional<ScheduleHeuristic> tryMerge(
   return SchedulerEntry::proposeHeuristics(fusion);
 }
 
+//! An utility class to compute and maintain the connection
+//!   relationship in a segmented graph. Space heavy and should
+//!   avoid using on very large graphs
+class BackwardReachability {
+  using GroupSet = std::unordered_set<SegmentedGroup*>;
+  using GroupSetPtr = std::unique_ptr<GroupSet>;
+  using ReachMap = std::unordered_map<SegmentedGroup*, GroupSetPtr>;
+
+ public:
+  //! Populate backward reachability map from segmented fusion
+  explicit BackwardReachability(SegmentedFusion* segmented_fusion)
+      : segmented_fusion_(segmented_fusion) {
+    computeReachability();
+  }
+
+  //! Does `g` have backward path to any group in `any`?
+  bool reachesAny(SegmentedGroup* g, const std::vector<SegmentedGroup*>& any) {
+    auto& reachable_from_g = getSet(g);
+    for (const auto& to : any) {
+      if (reachable_from_g->count(to)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  //! Update the map as the given two groups are now merged into ab
+  void mergeGroups(SegmentedGroup* a, SegmentedGroup* b, SegmentedGroup* ab) {
+    // TODO: avoid re-hashing
+    auto& ab_set = getSet(ab);
+    unionInto(ab, a);
+    unionInto(ab, b);
+    ab_set->erase(a);
+    ab_set->erase(b);
+    reach_map_.erase(a);
+    reach_map_.erase(b);
+    for (auto& it : reach_map_) {
+      if (it.second->count(a) || it.second->count(b)) {
+        it.second->insert(ab);
+        unionInto(it.first, ab);
+      }
+      it.second->erase(a);
+      it.second->erase(b);
+    }
+  }
+
+ private:
+  //! Collect initial reachable info using
+  //!  a work list algorithm through forward traversal
+  //!  a backward DFS would do the same
+  void computeReachability() {
+    GroupSet completed;
+    GroupSet work_list;
```

Comment:
Nit: I've typically been sticking with with `to_visit` and `visited`. Just makes them seem more related with that language.

Reply:
Updated. Thanks!